services/AttendanceService.ts

```diff
@@ -51,8 +51,36 @@ export default class AttendanceService {
       );
       await Repositories.user(txn).addPoints(user, pointsEarned);
 
-      const attendance = await attendanceRepository.attendEvent(user, event, attendedAsStaff);
-      return attendance.getPublicAttendance();
+      await attendanceRepository.attendEvent(user, event, attendedAsStaff);
+      return event.getPublicEvent();
+    });
+  }
+
+  public async attendEventRetroactively(userUuid: Uuid, eventUuid: Uuid): Promise<PublicEvent> {
+    return this.transactions.readWrite(async (txn) => {
+      const attendanceRepository = Repositories.attendance(txn);
+
+      const event = await Repositories.event(txn).findByUuid(eventUuid);
+      if (!event) throw new UserError('This event doesn\'t exist');
+      if (!event.hasStarted()) throw new UserError('This event hasn\'t started yet');
+
+      const user = await Repositories.user(txn).findByUuid(userUuid);
+      if (!user) throw new NotFoundError('This user was not found');
+
+      const hasAlreadyAttended = await attendanceRepository.hasUserAttendedEvent(user, event);
+      if (hasAlreadyAttended) throw new UserError('This user has already attended this event');
+
+      const pointsEarned = event.pointValue;
+      const attendedAsStaff = false;
```

Comment:
Why this constraint? Could an event host potentially want to check someone in as staff, e.g. they helped with pre-work/setup but couldn't attend the actual event?

Reply:
IMO this route shouldn't be used for things like pre-checkin / checking in as a staff if they didn't actually attend the event. This could bloat things like membership points or attendance head count if there's a bunch of people who weren't actually present but still counted for attendance. It's more meant for people who attended late, couldn't check, and who actually need the attendance as a requirement for something (e.g. Bit-Byte requires >= 2 attendances to bit-byte events). Either way, membership / ACM board should probably be the ones enforcing these constraints and not the backend so I'll remove them.